File: dlls/IVI Foundation/IVI/Drivers/wx218x/Examples/VC.NET/Cpp/ArbitraryWaveform/ArbitraryWaveform.cpp
// ArbitraryWaveform.cpp : Defines the entry point for the console application.

//Creates a driver object, reads a few Identity interface properties, and checks the instrument error queue.
//Additional instrument specific functionality:
//Defines 8 waveforms. Runs in simulation mode without an instrument.
//For WX2184 show how to download waveforms in Combine mode

#include "stdafx.h"
#define _USE_MATH_DEFINES
#include <cmath>

#import "IviDriverTypeLib.dll"  no_namespace
#import "IviFgenTypeLib.dll"  no_namespace
#import "wx218x_64.dll"  no_namespace


#pragma warning(disable : 4996)


SAFEARRAY * calculateSinusoidWaveform (long lPoints, double dCycles, double dStartPhase);

//Function that shows how to create buffer with combine waveform
unsigned long buildComposedWave(unsigned short* outBuff, unsigned long outBuffLen, unsigned long numCycles);

int _tmain(int argc, _TCHAR* argv[])
{
   ::CoInitialize(NULL);

   try
   {
      IWX218xPtr spDriver(__uuidof(WX218x));

      try
      {
         CString strModel;
         char strTmp[512] = "";
         int isSim = 1;  //variable for know user want to work in simulation mode or not
         long segSize = 1024;
         long segSizeCmb;
           _bstr_t strFileName1, strFileName2;

         long Handle1, Handle2;

         // IIviDriverIdentity properties - Initialize not required
         _bstr_t bstrIdentifier = spDriver->Identity->Identifier;
         wprintf(L"Identifier: %s\n", bstrIdentifier.GetBSTR());

         _bstr_t bstrRevision = spDriver->Identity->Revision;
         wprintf(L"Revision: %s\n", bstrRevision.GetBSTR());

         _bstr_t bstrVendor = spDriver->Identity->Vendor;
         wprintf(L"Vendor: %s\n", bstrVendor.GetBSTR());

         _bstr_t bstrDescription = spDriver->Identity->Description;
         wprintf(L"Description: %s\n", bstrDescription.GetBSTR());

         // Setup VISA resource descriptor
         //CString strResourceDesc = "GPIB0::4::INSTR";
         CString strResourceDesc = "TCPIP::192.168.0.180::5025::SOCKET";
         //CString strResourceDesc = "USB0::0x168B::0x2184::0000214901::INSTR";
         
         //CString strResourceDesc = "GPIB0::4::INSTR";

         // Setup IVI-defined initialization options
         CString strStandardInitOptions = 
            "Cache=true, InterchangeCheck=false, QueryInstrStatus=false, RangeCheck=true, RecordCoercions=false, Simulate=true";

         //CString strStandardInitOptions = 
           //"Cache=true, InterchangeCheck=false, QueryInstrStatus=false, RangeCheck=true, RecordCoercions=false, Simulate=true,DriverSetup=Trace=True,TraceArray=true";

          printf("Arbitrary Waveform Example by Tabor Electronics Ltd.\n");

          //printf("To run example in simulation mode, press number <1>, to connect, press <2>.\n");
          //scanf("%d", &isSim);

          /*if (isSim != 1)  //user want to connect to the instrument
          {
            printf("Enter Resource Name [TCPIP::192.168.0.187::5025::SOCKET]:\n "); scanf("%s", &strTmp);

             strResourceDesc = strTmp;

             strStandardInitOptions = 
            "Cache=true, InterchangeCheck=false, QueryInstrStatus=false, RangeCheck=true, RecordCoercions=false, Simulate=false";
          }//end if, user want to connect to the instrument*/


         
         spDriver->Initialize(LPCTSTR(strResourceDesc), VARIANT_TRUE, VARIANT_TRUE, LPCTSTR(strStandardInitOptions));

         // IIviDriverIdentity properties - Initialize required
         _bstr_t bstrInstModel = spDriver->Identity->InstrumentModel;
         wprintf(L"InstrumentModel: %s\n", bstrInstModel.GetBSTR());

         //Saving Model in strModel
         strModel = bstrInstModel.GetBSTR();

         _bstr_t bstrInstFirmwareRevision = spDriver->Identity->InstrumentFirmwareRevision;
         wprintf(L"InstrumentFirmwareRevision: %s\n", bstrInstFirmwareRevision.GetBSTR());

         _bstr_t bstrInstManufacturer = spDriver->Identity->InstrumentManufacturer;
         wprintf(L"InstrumentManufacturer: %s\n\n", bstrInstManufacturer.GetBSTR());

         /* Create a arbitrary waveform and load it to the instrument */
         int handles[8]; //array that containts waveform handles
         int sizeWaves[8];  //array that containts size for creating waveforms
         long hr = 0;

         _bstr_t bstrChannel = _T("Channel1");

         //Insert size in array sizeWaves
         sizeWaves[0] = 384;
         sizeWaves[1] = 4096;
         sizeWaves[2] = 81920;
         sizeWaves[3] = 10240;
         sizeWaves[4] = 384;
         sizeWaves[5] = 4096;
         sizeWaves[6] = 40960;
         sizeWaves[7] =  204800; 

         if (!strModel.CompareNoCase(_T("WX2184")) || !strModel.CompareNoCase(_T("WX1284")) || !strModel.CompareNoCase(_T("WX2184C")) || !strModel.CompareNoCase(_T("WX1284C"))) //it is WX2184
         {
            for (int cc = 0; cc < 4;)  //Loop for channels 1 and 3, becouse default mode is Duplicate
           {
            /* set a active select */
            spDriver->ActiveChannel = bstrChannel;

            /* set a Sample Clock */
            spDriver->Arbitrary->SampleRate = 75E6;
            
            /* assign the Arbitrary Mode */
            spDriver->Output->OutputMode = WX218xOutputModeArbitrary;

            /* turn on the outputs */
            spDriver->Output->Enabled[bstrChannel] = VARIANT_TRUE;

            spDriver->Output->Sync->Enabled[bstrChannel] = VARIANT_TRUE; 

            //Setting trace mode to duplicate
            spDriver->Arbitrary->Waveform->put_TraceMode(WX218xTraceModeDuplicate);

           for (int i = 0; i < sizeof(handles)/sizeof(handles[0]); i++) 
            {
               SAFEARRAY *pData = calculateSinusoidWaveform (sizeWaves[i], 1, 0);

               handles[i] = spDriver->Arbitrary->Waveform->Create (&pData); 
               SafeArrayDestroy (pData);

               wprintf(L"Download Wave for %s, segment %d\n\n", bstrChannel.GetBSTR(), i+1);
            }

            bstrChannel = _T("Channel3");

            cc += 2;

         }//end for, Loop for channels

         }//it is WX2184

         else //it is other models, not WX2184
         {

         for (int cc = 0; cc < 2; cc++)  //Loop for channels
         {
            /* set a active select */
            spDriver->ActiveChannel = bstrChannel;

            if (!strModel.CompareNoCase(_T("WX2184")) || !strModel.CompareNoCase(_T("WX2184C"))) //it is WX2184
            {
                spDriver->Arbitrary->SampleRate = 1E9;
            }

            else //it is not WX2184
            {
              /* set a Sample Clock */
              spDriver->Arbitrary->SampleRate = 10E6;
            }

            /* assign the Arbitrary Mode */
            spDriver->Output->OutputMode = WX218xOutputModeArbitrary;

            /* turn on the outputs */
            spDriver->Output->Enabled[bstrChannel] = VARIANT_TRUE;

            spDriver->Output->Sync->Enabled[bstrChannel] = VARIANT_TRUE; 

           

            for (int i = 0; i < sizeof(handles)/sizeof(handles[0]); i++) 
            {
               SAFEARRAY *pData = calculateSinusoidWaveform (sizeWaves[i], 1, 0);

               handles[i] = spDriver->Arbitrary->Waveform->Create (&pData); 
               SafeArrayDestroy (pData);

               wprintf(L"Download Wave for %s, segment %d\n\n", bstrChannel.GetBSTR(), i+1);
            }

            /* walk through the segments */
            for (int i = 0; i < sizeof(handles)/sizeof(handles[0]); i++) 
            {
               spDriver->Arbitrary->Waveform->Configure (bstrChannel, handles[i], 0.5, 0.0);
            }

            bstrChannel = _T("Channel2");

         }//end for, Loop for channels

         }//end else, it is not WX2184

         //Now checking if model is WX2184, show download combine waveform
          if (!strModel.CompareNoCase(_T("WX2184")) || !strModel.CompareNoCase(_T("WX1284")) || !strModel.CompareNoCase(_T("WX2184C")) || !strModel.CompareNoCase(_T("WX1284C"))) //it is WX2184
          {
            printf("Now creating combine waveform (Sine and Triangle) and download in to the instrument.\n");
            unsigned short pointCmb;
            long handle;

            bstrChannel = _T("Channel1");

            segSize = 1024;
            segSizeCmb = segSize*2 + 32;

            //Allocation buffer for combine waveform
            unsigned short * buffCombine = new unsigned short[segSizeCmb];

            int retLenCmb;   //size of waveform for one channel that returns from function: "buildComposedWave"

            retLenCmb = buildComposedWave(buffCombine, segSizeCmb, 3);

            //Now in buffer buffCombine exist data for combined wavefor
            //Creating Safearray for copy data
             SAFEARRAY *m_pCmbData = SafeArrayCreateVector(VT_I2, 0, segSizeCmb);

             for (long ii = 0; ii < segSizeCmb; ii++)
             {
              //Setting data from buffCombine in to the safearray
                pointCmb = buffCombine[ii]; 
                SafeArrayPutElement(m_pCmbData, &ii, (void*)&pointCmb);
             }//end for

            for (int cc = 0; cc < 4;)
            {
              /* set a active select */
             spDriver->ActiveChannel = bstrChannel;

              //Setting mode to Combine
             spDriver->Arbitrary->Waveform->TraceMode = WX218xTraceModeCombine;

             //Now download combined waveform in to the instrument
             handle = spDriver->Arbitrary->Waveform->CreateCustom(&m_pCmbData);
             wprintf(L"Download Wave for %s, Waveform Handle is %d\n\n", bstrChannel.GetBSTR(), handle);

             bstrChannel = _T("Channel3");
             cc += 2;

            }//end for

           delete [] buffCombine;

           SafeArrayDestroy (m_pCmbData);

           bstrChannel = _T("Channel1");
           spDriver->Output->Enabled[bstrChannel] = VARIANT_TRUE;
           bstrChannel = _T("Channel2");
           spDriver->Output->Enabled[bstrChannel] = VARIANT_TRUE;
           bstrChannel = _T("Channel3");
           spDriver->Output->Enabled[bstrChannel] = VARIANT_TRUE;
           bstrChannel = _T("Channel4");
           spDriver->Output->Enabled[bstrChannel] = VARIANT_TRUE;

           strFileName1 = _T("waves\\pulse_2048.wav");
           strFileName2 = _T("waves\\sinc_2048.wav");

           printf("Now download two waveforms from the binary files (Pulse and Sinc) in to the instrument.\n");
          
           bstrChannel = _T("Channel1");
           for (int cc = 0; cc < 4;)
           {
              /* set a active select */
             spDriver->ActiveChannel = bstrChannel;

              //Setting mode to Combine
             spDriver->Arbitrary->Waveform->TraceMode = WX218xTraceModeCombine;

             //Now download combined waveform in to the instrument
             hr = spDriver->Arbitrary->Waveform->LoadArbWfmFromFileAdv(bstrChannel, strFileName1, strFileName2, &Handle1, &Handle2);
             wprintf(L"Download two Waveforms, Waveform Handles are: %d,%d\n\n", Handle1, Handle2);

             bstrChannel = _T("Channel3");
             cc += 2;

           }//end for
           

         }//end if, it is WX2184


         // Check instrument for errors
         long lErrorNum = -1;
         _bstr_t bstrErrorMsg;
         wprintf(L"\n");
         while (lErrorNum != 0)
         {
            spDriver->Utility->ErrorQuery(&lErrorNum, bstrErrorMsg.GetAddress());
            wprintf(L"ErrorQuery: %d, %s\n", lErrorNum, bstrErrorMsg.GetBSTR());
         }

      }  //end try

      catch (_com_error& e) 
      {
         //::MessageBox(NULL, e.Description(), e.ErrorMessage(), MB_ICONERROR);  //open this row if you want to display message box
         long hr = e.Error();
         _bstr_t msg = e.Description();

         wprintf(L"ErrorCode: %d, %s\n", hr, msg.GetBSTR());
      }

      if (spDriver != NULL && spDriver->Initialized)
      {
         // Close driver
         spDriver->Close();
      }
   }
   catch (_com_error& e)
   {
      ::MessageBox(NULL, e.Description(), e.ErrorMessage(), MB_ICONERROR);
   }

   ::CoUninitialize();

   printf("\n\nDone - Press Enter to Exit");
   fflush(stdin);

   getchar();

   return 0;
}


/*
*  calculateSinusoidWaveform -
*/
SAFEARRAY * calculateSinusoidWaveform (long lPoints, double dCycles, double dStartPhase)
{
   double omega, point, phase, cycle;

   cycle = (double)(lPoints) / dCycles;

   /* Calculate a start phase */
   phase = (double)((double)(dStartPhase) * M_PI / (double)180.0);

   /* Calculate a Omega */
   omega = (double)(1.0 / lPoints * M_PI * 2 * dCycles);

   SAFEARRAY *m_pData = SafeArrayCreateVector(VT_R8, 0, lPoints);

   for (long i = 0; i < lPoints; i++) {
      point = sin ((omega * (double)i) + phase);
      SafeArrayPutElement(m_pData, &i, (void*)&point);
   }

   return m_pData;
}

//=====================================================================================================
//                   - How to create buffer with combine waveform
unsigned long buildComposedWave(unsigned short* outBuff, unsigned long outBuffLen, unsigned long numCycles)
{
      //Function that creates sine and triangle (combine waveform).
      double omg, val1, val2;
      unsigned long i, j, n;
      unsigned short *ptr1, *ptr2;

      const unsigned short dacMaxAbsVal = (0x01 << 14) - 1;
      const double dacHalfVal = dacMaxAbsVal / 2.0;
      
      unsigned long segLen = (outBuffLen / 32) * 16;
      if (segLen <= 16 + 192) { return 0; }
      segLen -= 16;

      ptr2 = &outBuff[32];
      ptr1 = ptr2 + 16;
      omg = 2 * M_PI / (segLen / (double)numCycles);
      n = 0;
      for (i = 0; i < segLen / 16; ++i, ptr1 += 16, ptr2 += 16)
      {
            for (j = 0; j < 16; ++j, ++n, ++ptr1, ++ptr2)
            {
                  val1 = sin(n * omg);
                  val2 = fabs(asin(val1) / M_PI) * dacMaxAbsVal;
                  val1 = (1.0 + sin(n * omg)) * dacHalfVal;
                  if (val1 < 0) { val1 = 0; }
                  if (val1 > dacMaxAbsVal) { val1 = dacMaxAbsVal; }
                  *ptr1 = (unsigned short)val1;
                  *ptr2 = (unsigned short)val2;
            }
      }

      for (i = 0; i < 16; ++i)
      {
            outBuff[i] = outBuff[32];
            outBuff[i + 16] = outBuff[48];
      }

      return segLen;
}

File: dlls/IVI Foundation/IVI/Drivers/wx218x/Examples/VC.NET/CSharp/ArbitraryWaveform/ArbitraryWaveform.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Text;
using Ivi.Driver.Interop;
using Ivi.Fgen.Interop;
using TaborElectronics.WX218x.Interop;
using Ivi.SessionFactory.Interop;

namespace ArbitraryWaveformCs
{
    /// <summary>
    /// Tabor Electronics IVI-COM Driver Example Program
    /// 
    /// Creates a driver object, reads a few Identity interface
    /// properties, and checks the instrument error queue.
    /// Additional instrument specific functionality:
    /// 
    /// Defines 8 waveforms.
    /// 
    /// Runs in simulation mode without an instrument.
    /// 
    /// Requires a COM reference to the driver's type library.
    /// 
    /// </summary>

    class ArbitraryWaveform
    {
        private static double[] mBlock = null;
        private const int MAX_ARBITRARY_MEMORY_SIZE = 64000000;

        [STAThread]
        public static void Main(string[] args)
        
        {
            WX218x driver = null;

            try
            {
                mBlock = new double[MAX_ARBITRARY_MEMORY_SIZE];

                driver = new WX218x();
                
                // IIviDriverIdentity properties - Initialize not required
                string identifier = driver.Identity.Identifier;
                Console.WriteLine("Identifier: {0}", identifier);

                string revision = driver.Identity.Revision;
                Console.WriteLine("Revision: {0}", revision);

                string vendor = driver.Identity.Vendor;
                Console.WriteLine("Vendor: {0}", vendor);

      
                // Setup VISA resource descriptor
                //string resourceDesc = "GPIB0::4::INSTR";
                string resourceDesc = "TCPIP::192.168.0.178::5025::SOCKET";
               // string resourceDesc = "USB0::0x168B::0x2184::2147516420::INSTR";


                // Setup IVI-defined initialization options
                string standardInitOptions = "Cache=false, InterchangeCheck=false, QueryInstrStatus=false, RangeCheck=true, RecordCoercions=false, Simulate=true";

                // Setup driver-specific initialization options
                //string driverSetupOptions = "DriverSetup = Model=WX2182, Simulate = true";
                //string driverSetupOptions = "DriverSetup = Model=WS8352, Simulate = false";
                string driverSetupOptions = "";

                //long numTmp = 1;   //variable for know that user want to work in simulation

               // TextReader tIn = Console.In;

                //Console.WriteLine("Arbitrary Waveform Example by Tabor Electronics Ltd");

                //Console.WriteLine("To run example in simulation mode, press number <1>, to connect, press <2>.");

               // numTmp = int.Parse(tIn.ReadLine());
                

                //Console.WriteLine("Enter Resource Name [TCPIP::192.168.0.180::5025::SOCKET]:");

                //resourceDesc = Console.ReadLine();

                //if (resourceDesc == "")
                 //   resourceDesc = "TCPIP::192.168.0.188::5025::SOCKET";


                if (standardInitOptions.Length > 0)
                    driverSetupOptions = standardInitOptions + "," + driverSetupOptions;
                driver.Initialize(resourceDesc, true, true, driverSetupOptions);

                // IIviDriverIdentity properties - Initialize required
                string instModel = driver.Identity.InstrumentModel;
                Console.WriteLine("InstrumentModel: {0}", instModel);

                string instFirmwareRevision = driver.Identity.InstrumentFirmwareRevision;
                Console.WriteLine("InstrumentFirmwareRevision: {0}", instFirmwareRevision);

                string instManufacturer = driver.Identity.InstrumentManufacturer;
                Console.WriteLine("InstrumentManufacturer: {0}", instManufacturer);

                // Create a arbitrary waveform and load it to the instrument
                int[] handles = new int[8];
                int[] sizeWaves = new int[8];  //array that containts size for creating waveforms
                string Channel = "Channel1", strCommand;

                //Insert size in array sizeWaves
                sizeWaves[0] = 384;
                sizeWaves[1] = 4096;
                sizeWaves[2] = 81920;
                sizeWaves[3] = 10240;
                sizeWaves[4] = 384;
                sizeWaves[5] = 4096;
                sizeWaves[6] = 40960;
                sizeWaves[7] = 204800;

                if (instModel == "WX2184" || instModel == "WX1284" || instModel == "WX2184C" || instModel == "WX1284C") //it is WX2184
                {
                    for (int cc = 0; cc < 4; )  //Loop for two channels (Channel1 and Channel3, becouse default mode is Duplicate)
                    {
                        // Set a active select
                        driver.ActiveChannel = Channel;

                        // Set a arbitrary mode
                        driver.Output.OutputMode = WX218xOutputModeEnum.WX218xOutputModeArbitrary;

                        // Set a sample clock
                        driver.Arbitrary.SampleRate = 75E6;


                        // Turn on the outputs
                        driver.Output.set_Enabled(Channel, true);

                        //Turn on the sync
                        driver.Output.Sync.set_Enabled(Channel, true);

                        //Setting trace mode to duplicate
                        driver.Arbitrary.Waveform.TraceMode = WX218xTraceModeEnum.WX218xTraceModeDuplicate;

                        for (int i = 0; i < handles.Length; i++)
                        {
                            //Creating array of data values (short) in range 0 to 4095
                            short[] data = ConvertWaveformDoubleToDAC(calculateSinusoidWaveform(sizeWaves[i], 1, 0), sizeWaves[i], instModel);

                            //Download wave in to the instrument
                            handles[i] = driver.Arbitrary.Waveform.CreateCustom(ref data);

                            strCommand = String.Format("Download wave for {0}, segment {1}, length {2}\n", Channel, i + 1, sizeWaves[i]);
                            Console.WriteLine(strCommand);

                            //free system memory
                            GC.Collect();
                            GC.WaitForPendingFinalizers();
                        }

                        Channel = "Channel3";
                        cc += 2;
                        

                    }//end for Loop
                }//end if, it is WX2184

                else //another models, not WX2184
                {

                    for (int cc = 0; cc < 2; cc++)  //Loop for two channels
                    {
                        // Set a active select
                        driver.ActiveChannel = Channel;

                        // Set a arbitrary mode
                        driver.Output.OutputMode = WX218xOutputModeEnum.WX218xOutputModeArbitrary;

                        // Set a sample clock
                        driver.Arbitrary.SampleRate = 10E6;


                        // Turn on the outputs
                        driver.Output.set_Enabled(Channel, true);

                        //Turn on the sync
                        driver.Output.Sync.set_Enabled(Channel, true);

                        for (int i = 0; i < handles.Length; i++)
                        {
                            //Creating array of data values (short) in range 0 to 4095
                            short[] data = ConvertWaveformDoubleToDAC(calculateSinusoidWaveform(sizeWaves[i], 1, 0), sizeWaves[i], instModel);

                            //Download wave in to the instrument
                            handles[i] = driver.Arbitrary.Waveform.CreateCustom(ref data);

                            strCommand = String.Format("Download wave for {0}, segment {1}, length {2}\n", Channel, i + 1, sizeWaves[i]);
                            Console.WriteLine(strCommand);

                            //free system memory
                            GC.Collect();
                            GC.WaitForPendingFinalizers();
                        }


                        //walk through the segments 
                        for (int i = 0; i < handles.Length; i++)
                        {
                            driver.Arbitrary.Waveform.Configure(Channel, handles[i], 0.5, 0.0);
                        }

                        Channel = "Channel2";

                    }//end for Loop

                }//end else, another models, not WX2184

                // Check instrument for errors
                int errorNum = -1;
                string errorMsg = null;
                Console.WriteLine("");
                while (errorNum != 0)
                {
                    driver.Utility.ErrorQuery(ref errorNum, ref errorMsg);
                    Console.WriteLine("ErrorQuery: {0}, {1}", errorNum, errorMsg);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            finally
            {
                if (driver != null && driver.Initialized)
                {
                    // Close driver
                    driver.Close();
                }
            }

            Console.WriteLine("Done - Press Enter to Exit");
            Console.ReadLine();
        }

        /**
         * calculateSinusoidWaveform - 
         */
        static double[] calculateSinusoidWaveform(long lPoints, double dCycles, double dStartPhase)
        {
            double omega, phase, cycle;

            cycle = (double)(lPoints) / dCycles;

            /* Calculate a start phase */
            phase = (double)((double)(dStartPhase) * Math.PI / (double)180.0);

            /* Calculate a Omega */
            omega = (double)(1.0 / lPoints * Math.PI * 2 * dCycles);

            for (long i = 0; i < lPoints; i++)
                mBlock[i] = Math.Sin((omega * (double)i) + phase);

            return mBlock;
        }

        /*
         *  ConvertWaveformDoubleToDAC -
         */
        static short[] ConvertWaveformDoubleToDAC(double[] data, int size, string model)
        {
            //Function that recieve array of data between -1 and 1 and 
            //convert to DAC and load in to the memory and return waveform handle
            const double HALF_RANGE_12BIT = 2047.5;
            const short OFFSET_12BIT = 2048;
            const double HALF_RANGE_14BIT = 4094.5;
            const short OFFSET_14BIT = 4095;
            double half_range_current = 2047.5;
            short offset_current = 2048;

            if (model == "WX2181B" || model == "WX2182B" || model == "WX2181C" || model == "WX2182C" || model == "WX1281C" || model == "WX1282C" || model == "WX2184" || model == "WX1284" || model == "WX2184C" || model == "WX1284C") //it is 14bit
            {
                half_range_current = HALF_RANGE_14BIT;
                offset_current = OFFSET_14BIT;
            }//end if,it is 14bit

            else
            {
                half_range_current = HALF_RANGE_12BIT;
                offset_current = OFFSET_12BIT;
            }

            short[] block = new short[size];

            for (long i = 0; i < size; i++)
            {
                double point = data[i];

                //Checking that every element in array data must be between -1 to +1
                if (point < -1.0 || point > 1.0) //invalid value in waveform data
                {
                    return null;
                }//end if,invalid value in waveform data 

                block[i] = (short)((half_range_current * point) + offset_current);
            }

            return block;
        }
    }
}
